ddtrace/debugging/_session.py

```diff
@@ -0,0 +1,34 @@
+from dataclasses import dataclass
+import typing as t
+from weakref import WeakKeyDictionary as wkdict
+
+from ddtrace import tracer
+
+
+@dataclass
+class Session:
+    ident: str
+    level: int
+
+    def link_to_trace(self, trace_context: t.Optional[t.Any] = None):
+        SessionManager.link_session_to_trace(self, trace_context)
+
+    @classmethod
+    def get_for_trace(cls) -> t.Optional["Session"]:
+        return SessionManager.get_session_for_trace()
+
+
+class SessionManager:
+    _session_trace_map: wkdict = wkdict()  # Trace context to Session mapping
+
+    @classmethod
+    def link_session_to_trace(cls, session, trace_context: t.Optional[t.Any] = None) -> None:
+        cls._session_trace_map[trace_context or tracer.current_root_span()] = session
```

Comment:
passing trace_context can break the get_session_for_trace as it only works for the root_spans. why do we want to let user specify this?

Reply:
Unfortunately the tracer propagates headers by activating a context object, which is not retrievable with the current tracer API (i.e. the root span of a distributed trace does not carry the distributed header information, which is attached to the context object that is used as a parent for the local root)